### PR TITLE
feat: add resourcegen cmd

### PR DESCRIPTION
**Pull request by hugorut**
Adds resource generator cmd, allows for resource files to be built using a single command:

---

I started playing around with this idea when I was working on the last resource. Thought it might be a good addition to help out with contributions and make everyones lives easier. That being said this is a POC at the moment (only works with terraform) as I wanted to get everyones feedback to see if this actually worthwhile merging. I'll outline my thinking below.

## Changes

This PR adds a new `resourcegen` cmd which can be built or simply run with `go run` it takes a few flags to help with generating resources:

```
  -cloud_provider string
    	cloud provider to create resource for, one of [aws, azure, google] (default "aws")
  -resource_name string
    	the resource name to generate, use underscores between names, e.g. autoscaling_group
  -with_help
    	generate your resources with doc blocks and examples to help you get started. Useful for understanding how to add a resource. (default true)
```

The command then generates all the necessary files using go templating to replace all the relevant `var names`, `filenames`,  `directory names` and file contents so that it's relevant to the resource_name and provider passed. It also edits the resource registry to **add the correct get resource function**.

`resourcegen` also comes with a `with_help` flag, that, if set, adds a number of verbose comment and helper notes for starting contributors.

[Here's a gist with all the generated files with `with_help` set to true](https://gist.github.com/hugorut/d1c10ec4b61558a11122f0f8a9b200ff) (see the first image for the directory structure generated. The filenames in the gist don't correspond to the image because of naming collisions in the gist, but you get the idea.)

For everyday contributors who don't want just a skeleton they can run the cmd with `with_help=false` and just generate the bare bones.

[Here's a gist with the example output with `with_help` set to false](https://gist.github.com/hugorut/bd4935b84722175c34f4adcd05975730)

### Pros

* Speeds up adding resource boilerplate
* Standardises new resources, contributors have a single point to generate and can't misunderstand things and go rogue with file placement and naming
* Simplifies contributing docs and process - reference running the cmd and then follow particular steps afterwards
* Who doesn't like CLI tools

### Cons

* Increases code maintenance - if we change resources definitions and methods we need to also be sure to update the `resourcegen` tool so it doesn't degrade
* More stuff to understand

---

As mentioned at the start - the cmd currently just works for terraform. It would be an easy change to get it working with other providers but I wanted to hold off on doing that until we have a discussion about if we actually want it. I won't be offended if you decide it's too much, it was fun tinkering on this for a few hours this week.

### CODE REVIEW

**Comment by sinabakh** [comment]
@hugorut Awesome!
I've been personally thinking about how can we ease the resource addition process and the code-gen approach seems to be the answer. I believe this PR could be a good start for efforts leading to easier resource addition.

Apart from the pros and cons, you mentioned I can think of a few other points:

1. This approach will add this cmd to the main infracost executable, which is mostly used not for development purposes. I think we can have a different executable for such a tool like `infracost-gen` which can have its own set of dependencies in the future.
2. I personally think the final version of this generator needs more than a few arguments so I lean toward a more interactive approach instead of arguments. Like asking the resource name, then asking about IAC and etc...

Apart from the aforementioned points, I like how this eliminates the hassle of creating multiple files.

**Comment by hugorut** [comment]
@sinabakh glorious!

> 1. This approach will add this cmd to the main infracost executable, which is mostly used not for development purposes. I think we can have a different executable for such a tool like `infracost-gen` which can have its own set of dependencies in the future.

I may be misunderstanding you, but this approach doesn't add it to the main infracost executable as it's in a separate folder, so it's built as a completely different binary. If you mean it adds the **dependencies for the codegen tool** into the project then yes I agree. I did think about moving this into a separate repo. However, I think the benefits of having the codegen tool close to the code it's actually generating is a big pro keeping them together.

> 2. I personally think the final version of this generator needs more than a few arguments so I lean toward a more interactive approach instead of arguments. Like asking the resource name, then asking about IAC and etc...

Yup completely agree with this, but as you say I think this is a more polished version down the line. This is quite rough and ready at the moment. In the future it would be awesome to have arguments like `cost components` and `usage params` to build a skeleton right out the box. 

Happy to move this to a discussions thread if this is a better place to discuss - alternatively slack might be a better shout. Just thought it would be good to show a working concept as a good starting point for discussion.

**Comment by sinabakh** [comment]
@hugorut 
1. My bad, I misread the code. It's just the way I meant.
2. Sure, I was just a giving some ideas.

I would like to also know what others think about such approach to resource gen.

**Comment by hugorut** [comment]
@sinabakh I think the interactive builder approach could be really cool, plus it's easier to build up a more complex config. e.g. branching logic. So i'm 100% with you on it.

File: cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	//
+	// These properties are configured directly from attributes parsed in the IAC provider see
+	// your resource file at: internal/providers/terraform/{{ .CloudProvider }}/{{ .Filename }}.go
+	// for an example in how this is achieved.
+	//
+	// below are a few examples of common properties that are often part of resources:
+	{{- end }}
+	// Address is the unique name of the resource in the IAC language
+	Address       string
+	// Region is the aws region the {{ .ResourceName }} is provisioned within
```
Comment (vdmgolub) [inline cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl:31]: Guess we can also replace the provider here :)
```suggestion
	// Region is the {{ .CloudProvider }} region the {{ .ResourceName }} is provisioned within
```

**Review by vdmgolub — COMMENTED** [review]
@hugorut Great stuff!! I've left some minor comments, please take a look. Also please let me know if you need help with anything :)

File: cmd/resourcegen/main.go [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+}
+`, c.CloudProvider, c.CloudProvider)
+}
+
+func (c config) registryLocation() string {
+	return "internal/providers/terraform/" + c.CloudProvider + "/registry.go"
+}
+
+func (c config) registryFuncName() string {
+	return "get" + c.ResourceName + "RegistryItem"
```
[the remark this answers is not in the document]
Comment (hugorut) [reply cmd/resourcegen/main.go:127]: yeah I was kinda in two minds about this, currently these registry functions aren't used outside of the `aws` package so instinctively I went with the un-exported func. 

I had discussion with @aliscott about these functions in a prior PR adding a new resource. We decided at the time that making it un-exported was ok - see here https://github.com/infracost/infracost/pull/1064#discussion_r732494424. 

In my mind having them un-exported makes it a little easier to understand as people won't think they're used across the codebase. However, I really don't mind and I think the consistency argument is also a very good reason to keep them capitalised.

[the remark this answers is not in the document]
Comment (vdmgolub) [reply cmd/resourcegen/main.go:127]: Thanks for the context! Yep, I agree, the idiomatic Go code can be helpful on its own. Not sure what's the best way to make it default way though. We could update the contributing guide + all the existing `GetX` functions in one PR, but that might be a bit noisy :) Or update the guide only and make sure that all new resources use lowerCase functions.

For this PR let's keep it lower-cased then. :)

Also, I noticed in Tim's video, that we do our best effort to keep this functions list sorted alphabetically, however the complexity of the generator will increase to follow such practice. Maybe we can address it later 🤷‍♂️.

File: cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+// It uses the `infracost_usage` struct tags to populate data into the {{ .ResourceName }}.
+func ({{ .ResourceLetter }} *{{ .ResourceName }}) PopulateUsage(u *schema.UsageData) {
+	resources.PopulateArgsWithUsage({{ .ResourceLetter }}, u)
+}
+
+// BuildResource builds a schema.Resource from a valid {{ .ResourceName }}.
+// This method is called after the resource is initialised by an iac provider.
+// See providers folder for more information.
+func ({{ .ResourceLetter }}  *{{.ResourceName}}) BuildResource() *schema.Resource {
+	{{- if .WithHelp -}}
```
[the remark this answers is not in the document]
Comment (hugorut) [reply cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl]: resolved in https://github.com/infracost/infracost/pull/1091/commits/a2557a072da0c16e50c972570059b4959b79ff38

File: cmd/resourcegen/embed/providers/terraform/$cloud_provider$/testdata/$filename$_test/$filename$_test.tf.tmpl [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{{ .CloudProviderTerraformBlock }}
+
+# add the terraform for your {{ .ResourceName }} below
+# e.g:
+resource "{{ .CloudProvider }}_{{ .Filename }}" "{{ .Filename }}" {
+	aws_id = "fake"
+}
```
[the remark this answers is not in the document]
Comment (hugorut) [reply cmd/resourcegen/embed/providers/terraform/$cloud_provider$/testdata/$filename$_test/$filename$_test.tf.tmpl]: resolved in https://github.com/infracost/infracost/pull/1091/commits/a2557a072da0c16e50c972570059b4959b79ff38

File: cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+	"github.com/infracost/infracost/internal/resources"
+	"github.com/infracost/infracost/internal/schema"
+)
+
+// TODO: fill in the docs {{ .ResourceName }} with a short description of the cloud resource and
+// add any links to the pricing pages or docs that might be useful to developers in the future.
+// e.g.
+// 		more resource information here: https://aws.amazon.com/{{ .Filename }}/
+//
+//		pricing information here: https://aws.amazon.com/{{ .Filename }}/pricing/
```
[the remark this answers is not in the document]
Comment (hugorut) [reply cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl]: resolved in https://github.com/infracost/infracost/pull/1091/commits/a2557a072da0c16e50c972570059b4959b79ff38

File: cmd/resourcegen/embed/providers/terraform/$cloud_provider$/testdata/$filename$_test/$filename$_test.golden.tmpl [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{{- if .WithHelp -}}
+# add your expected infracost output for {{ .ResourceName }} below
+# test output can be populated into this file by running the
+# following cmd from the root directory:
```
[the remark this answers is not in the document]
Comment (hugorut) [reply cmd/resourcegen/embed/providers/terraform/$cloud_provider$/testdata/$filename$_test/$filename$_test.golden.tmpl]: fixed up comment punctuation and capitalizing in https://github.com/infracost/infracost/pull/1091/commits/a2557a072da0c16e50c972570059b4959b79ff38

File: cmd/resourcegen/main.go [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+func exitWithErr(err error) {
+	fmt.Fprint(os.Stderr, err.Error()+"\n")
+	os.Exit(1)
+}
+
+func writeOutput(c config, written []string) {
+	b := strings.Builder{}
+	b.WriteString(fmt.Sprintf("\nsuccessfully generated resource %s, files written:\n\n", c.Filename))
```
[the remark this answers is not in the document]
Comment (hugorut) [reply cmd/resourcegen/main.go]: fixed in https://github.com/infracost/infracost/pull/1091/commits/a2557a072da0c16e50c972570059b4959b79ff38

File: cmd/resourcegen/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hugorut) [reply cmd/resourcegen/main.go:127]: Yeah sorting alphabetically might add some complexity here - shall we park it for this PR and maybe open a issue to track?

File: cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl (hunk 1 above)
Comment (hugorut) [reply cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl]: fixed in https://github.com/infracost/infracost/pull/1091/commits/a2557a072da0c16e50c972570059b4959b79ff38

**Comment by alikhajeh1** [comment]
This is great! I gave it a quick try on a new resource to test the interface a bit. Should we keep it consistent with the other infracost commands that use two dashes for flags and one dashe between the flag words? e.g. `--terraform-plan-flags`

File: cmd/resourcegen/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (vdmgolub) [reply cmd/resourcegen/main.go:127]: Sure thing, as we already have at least one lowercase function in master, we can continue using the new practice. Will be a good chore/techdebt to tackle later :)

File: cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	// for an example in how this is achieved.
+	//
+	// Below are a few examples of common properties that are often part of resources:
+	{{- end }}
+	// Address is the unique name of the resource in the IAC language.
+	Address       string
+	// Region is the {{ .CloudProvider }} region the {{ .ResourceName }} is provisioned within.
+	Region        string
+	{{- if .WithHelp }}
+	// InstanceCount is an example properties designed to show you a complete example with cost components.
```
Comment (vdmgolub) [inline cmd/resourcegen/embed/resources/$cloud_provider$/$filename$.go.tmpl:35]: Not sure if this comment should list all three properties, or just InstanceCount.
```suggestion
	// InstanceCount is an example property designed to show you a complete example with cost components.
```

**Review by vdmgolub — COMMENTED** [review]
Sorry for late catches, was running tests for my resource and noticed couple things. Nothing serious, we're almost there! :)

**Comment by hugorut** [comment]
> This is great! I gave it a quick try on a new resource to test the interface a bit. Should we keep it consistent with the other infracost commands that use two dashes for flags and one dashe between the flag words? e.g. `--terraform-plan-flags`

Yup good idea

File: cmd/resourcegen/main.go [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+	var c config
+	flag.StringVar(&c.CloudProvider, "cloud-provider", "aws", "Cloud provider to create resource for, one of [aws, azure, google]")
+	flag.StringVar(&c.Filename, "resource-name", "", "The resource name to generate, use underscores between names, e.g. autoscaling_group (required)")
+	flag.BoolVar(&c.WithHelp, "with-help", true, "Generate your resources with doc blocks and examples to help you get started. Useful for understanding how to add a resource")
+	flag.Parse()
+
+	if flag.NFlag() == 0 {
+		flag.PrintDefaults()
+		os.Exit(1)
+	}
```
Comment (vdmgolub) [inline cmd/resourcegen/main.go:48]: @hugorut Tried to address Ali's comment (I guess we can live with single dash flags for now without introducing cobra), but ended up adding some extra formatting changes (see last commit). Hope it helps, feel free to change or throw away. :)

Comment (hugorut) [reply cmd/resourcegen/main.go:48]: the go default flags package allows both `--` and `-` to define the flags so double dash flags should work out the box. See the example image in the PR description (has `--` flags). Changes lGTM 👍

Comment (vdmgolub) [reply cmd/resourcegen/main.go:48]: Yeah, sorry, I meant we don't need to enforce the double dashes. Thanks for review! :)

**Review by vdmgolub — COMMENTED** [review]
@hugorut I believe we are good to go. Please take a look at my last two edits and if they look okay, I'm going to merge the PR 🎉.

**Comment by hugorut** [comment]
@vdmgolub Thanks for the edits, LGTM !

**Pull request merged**